Apromore-Custom-Plugins/Log-Animation-Logic/src/main/java/org/apromore/service/loganimation/replay/Optimizer.java

```diff
@@ -1,166 +1,183 @@
+            if (!newTrace.isEmpty()) newLog.add(newTrace);
+        }
+        return newLog;
+    }
+    
+    /**
+     * Update a process definition to make every node containing a name ref
+     * with reference to the name value (stored in map field of LogOptimizer)
+     * @param definitions: original process definition
+     * @return process definition after being modified
+     */
+    public Definitions optimizeProcessModel(Definitions definitions) {        
+        List<BaseElement> rootElements = definitions.getRootElement();
+        Process process = (Process)rootElements.get(0);        
+        for (FlowElement element : process.getFlowElement()) {
+            if (element instanceof FlowNode) {
+                ((FlowNode)element).setNameRef(cacheObject(element.getName()));
+            }
+        }
+        return definitions;
+    }
+    
+    /**
+     * Copy an attribute to an element. Cache the attribute in memory if it is not yet cached.
+     * @param key: the key of the attribute
+     * @param attribute: the attribute to be copied
+     * @param newElement: the element which will contain a copy of the attribute
+     * @return true: the attribute has been copied, false otherwise.
+     */
+    private void copyCachedAttribute(String key, XAttribute attribute, XAttributable newElement) {
+        if (attribute == null || newElement == null || key == null || key.isEmpty()) return ;
+        Object currentAttValue = getAttributeValue(attribute);
+        if (currentAttValue != null) {
+            String cachedKey = (String) cacheObject(key); // cache key
+            Object cachedValue = cacheObject(currentAttValue); // cache value
+            XAttribute newAtt = createXAttribute(cachedKey, cachedValue, attribute);
+            newElement.getAttributes().put(cachedKey, newAtt);
+        }
+    }
+
+    /**
+     * Cache an object
+     * @param o: an object
+     * @return a reference to an already cached object with the same value as o, 
+     *          or a reference to o if no cache exists.
+     */
+    private Object cacheObject(Object o) {
+        Object result = null;
+        if(o instanceof Date) return o;
+        if((result = cache.get(o)) == null) {
+            cache.put(o, o);
+            result = o;
+        }
+        return result;
+    }
+
+    private Object getAttributeValue(XAttribute attribute) {
+        if(attribute instanceof XAttributeLiteral) return ((XAttributeLiteral) attribute).getValue();
+        else if(attribute instanceof XAttributeBoolean) return ((XAttributeBoolean) attribute).getValue();
+        else if(attribute instanceof XAttributeDiscrete) return ((XAttributeDiscrete) attribute).getValue();
+        else if(attribute instanceof XAttributeContinuous) return ((XAttributeContinuous) attribute).getValue();
+        else if(attribute instanceof XAttributeTimestamp) return ((XAttributeTimestamp) attribute).getValue();
+        else {
+            System.out.println(attribute);

```

Comment:
Would it be better to use LOGGER?

Reply:
I was so distracting trying to figure out whether there was a way to avoid all that typecasting in the preceding lines that I completely missed that!

Reply:
Yeah, this section of code was written by the original author, I didn't look at it. It has some issues with the error handling is not proper  when the attribute is not one of the accepted types, only printing out with System.out.

The casting is one thing needs to do because of OpenXES's design, perhaps an issue, XAttribute doesn't have getValue() method in its interface, while getValue() is present in other extended interface such as XAttributeDiscrete, XAttributeTimestamp, etc.

Reply:
We'll have to pay attention to these details in the new implementation.